cmd/util/cmd/read-protocol-state/cmd/snapshot.go

```diff
@@ -0,0 +1,83 @@
+package cmd
+
+import (
+	"github.com/rs/zerolog/log"
+	"github.com/spf13/cobra"
+
+	"github.com/onflow/flow-go/cmd/util/cmd/common"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/state/protocol/inmem"
+)
+
+var SnapshotCmd = &cobra.Command{
+	Use:   "snapshot",
+	Short: "Read snapshot from protocol state",
+	Run:   runSnapshot,
+}
+
+func init() {
+	rootCmd.AddCommand(SnapshotCmd)
+
+	SnapshotCmd.Flags().Uint64Var(&flagHeight, "height", 0,
+		"Block height")
+
+	SnapshotCmd.Flags().BoolVar(&flagFinal, "final", false,
+		"get finalized block")
+
+	SnapshotCmd.Flags().BoolVar(&flagSealed, "sealed", false,
+		"get sealed block")
+}
+
+func runSnapshot(*cobra.Command, []string) {
+	db := common.InitStorage(flagDatadir)
+	defer db.Close()
+
+	storages := common.InitStorages(db)
+	state, err := common.InitProtocolState(db, storages)
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not init protocol state")
+	}
+
+	var snapshot protocol.Snapshot
+
+	if flagHeight > 0 {
+		log.Info().Msgf("get snapshot by height: %v", flagHeight)
+		snapshot = state.AtHeight(flagHeight)
+	}
+
+	if flagFinal {
+		log.Info().Msgf("get last finalized snapshot")
+		snapshot = state.Final()
+	}
+
+	if flagSealed {
+		log.Info().Msgf("get last sealed snapshot")
+		snapshot = state.Sealed()
+	}
+
+	head, err := snapshot.Head()
+	if err != nil {
+		log.Fatal().Err(err).Msg("fail to get block of snapshot")
+	}
+
+	log.Info().Msgf("creating snapshot for block height %v, id %v", head.Height, head.ID())
+
+	serializable, err := inmem.FromSnapshot(snapshot)
+	if err != nil {
+		log.Fatal().Err(err).Msg("fail to serialize snapshot")
+	}
+
+	sealingSegment, err := serializable.SealingSegment()
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not get sealing segment")
+	}
+
+	log.Info().Msgf("snapshot created, sealed height %v, id %v",
+		sealingSegment.Sealed().Header.Height, sealingSegment.Sealed().Header.ID())
+
+	log.Info().Msgf("highest finalized height %v, id %v",
+		sealingSegment.Highest().Header.Height, sealingSegment.Highest().Header.ID())
+
+	encoded := serializable.Encodable()
+	common.PrettyPrint(encoded)
```

Comment:
nitpick: the command is combining informational logging with printing the desired output. I think this means we need to manually copy-paste only a portion of the output from the terminal. And the output is generally quite large. 

It's better to either:
- log as much as you want, but write the desired output to a file
- don't log unless something has failed, only print desired output in happy-path, then we can do `util snapshot > snapshot.json`

Reply:
`common.PrettyPrint` will print to standard out, where as log.Info will print to standard error. So it's actually quite clear to see the last sealed block ID when we forwarding the snapshot to a file:

```
execution-002 /var/flow # ./util read-protocol-state snapshot --final --datadir /var/flow/data/protocol > /var/flow/bootstrap2/public-root-information/root-protocol-state-snapshot.json
11:59PM INF get last finalized snapshot
11:59PM INF creating snapshot for block height 36329882, id c966de34fdb0b961376bc7a366c06c3a9dfca3468b97fbcd5db45ec9cf7557a2
11:59PM INF snapshot created, sealed height 36329876, id 11970583b31e6f43fb77ae09e48e0588ad3010c4514ac62468b837ce8cf1afff
11:59PM INF highest finalized height 36329882, id c966de34fdb0b961376bc7a366c06c3a9dfca3468b97fbcd5db45ec9cf7557a2
```

If we want to read the last sealed block ID from the snapshot, we could just load the json and read `LatestSeal.BlockID` field of the JSON object.

Reply:
Awesome - didn't know we already had it set up that way 👌